rustworkx-core/src/dag_algo.rs

```diff
@@ -313,6 +316,119 @@ where
     Ok(Some((path, path_weight)))
 }
 
+/// Return a list of graph layers
+///
+/// A layer is a subgraph whose nodes are disjoint, i.e.,
+/// a layer has depth 1. The layers are constructed using a greedy algorithm.
+///
+/// Arguments:
+///
+/// * `graph` - The graph to get the layers from
+/// * `first_layer` - A list of node ids for the first layer. This
+///     will be the first layer in the output
+///
+/// ```
+/// use rustworkx_core::petgraph::prelude::*;
+/// use rustworkx_core::dag_algo::layers;
+/// use rustworkx_core::dictmap::*;
+///
+/// let edge_list = vec![
+///  (0, 1),
+///  (1, 2),
+///  (2, 3),
+///  (3, 4),
+/// ];
+///
+/// let graph = DiGraph::<u32, u32>::from_edges(&edge_list);
+/// let layers: Vec<Vec<NodeIndex>> = layers(&graph, vec![0.into(),1.into()]).unwrap();
+/// let expected_layers: Vec<Vec<NodeIndex>> = vec![
+///     vec![0.into(),1.into()],
+///     vec![1.into(),2.into()],
+///     vec![2.into(),3.into()],
+///     vec![3.into(),4.into()],
+///     vec![4.into()]
+/// ];
+/// assert_eq!(layers, expected_layers)
+/// ```
+pub fn layers<G>(graph: G, first_layer: Vec<G::NodeId>) -> Result<Vec<Vec<G::NodeId>>, LayersError>
+where
+    G: NodeIndexable // Used in from_index and to_index.
+        + NodeCount // Used in find_cycle
+        + EdgeCount // Used in find_cycle
+        + Visitable // Used in find_cycle
+        + IntoNodeIdentifiers // Used for .node_identifiers
+        + IntoNeighborsDirected // Used for .neighbors_directed
+        + IntoEdgesDirected, // Used for .edged_directed
+    <G as GraphBase>::NodeId: Eq + Hash,
+{
+    let mut output_indices: Vec<Vec<G::NodeId>> = Vec::new();
+    let mut cur_layer = first_layer;
+    let mut next_layer: Vec<G::NodeId> = Vec::new();
+    let mut predecessor_count: HashMap<G::NodeId, usize> = HashMap::new();
+    let node_set = graph.node_identifiers().collect::<HashSet<G::NodeId>>();
+    // Throw error if a cycle is found at the current node
+    if let Some(node) = &cur_layer.first() {
+        let check_cycle = find_cycle(&graph, Some(**node));
+        if !check_cycle.is_empty() {
+            return Err(LayersError(Some(
+                "The provided graph has a cycle".to_string(),
+            )));
+        }
+    }
+    for layer_node in &cur_layer {
+        if !node_set.contains(layer_node) {
+            return Err(LayersError(Some(format!(
+                "An index input in 'first_layer' {} is not a valid node index in the graph",
+                graph.to_index(*layer_node)
+            ))));
+        }
+    }
```

Comment:
Do we need this in the rust API? I'm wondering if we should just treat this with a panic because it's a straight programming error in rust, not an expected input. We need this python side so we avoid a panic and return an `IndexError` but in rust this condition is the same as doing:

```rust
let test = vec![0, 1]
println!("{}", test[42])
```

so guarding against it feels unnecessary.

Reply:
I'll move it to Python only then

Reply:
This check might still be needed as it is only performed on that first layer, which is not validated in any way. So I kept it, but a `panic!` is called instead. I wonder if this is not the desired behavior @mtreinish.